oauthproxy_test.go

```diff
@@ -161,6 +160,227 @@ func TestEncodedSlashes(t *testing.T) {
 	}
 }
 
+type fakeUpstreamSetup struct {
+	lastSeenURI     string
+	lastSeenRequest string
+	lastSeenCode    int
+	lastCode        int
+	proxy           *OAuthProxy
+	backend         *httptest.Server
+	backendURL      *url.URL
+}
+
+func newFakeUpstreamSetup(upstreamPath string) *fakeUpstreamSetup {
+
+	setup := &fakeUpstreamSetup{}
+
+	setup.backend = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		setup.lastSeenURI = r.RequestURI
+		if dump, err := httputil.DumpRequest(r, false); err != nil {
+			setup.lastSeenCode = 500
+			w.WriteHeader(500)
+			setup.lastSeenRequest = fmt.Sprintf("%s", err)
+		} else {
+			setup.lastSeenCode = 200
+			w.WriteHeader(200)
+			setup.lastSeenRequest = fmt.Sprintf("%s", dump)
+		}
+	}))
+
+	setup.backendURL, _ = url.Parse(setup.backend.URL)
+
+	opts := NewOptions()
+	opts.ClientID = "asdlkjx"
+	opts.ClientSecret = "alkgks"
+	opts.CookieSecret = "asdkugkj"
+	opts.SkipAuthRegex = []string{".*"} // don't want auth here
+	opts.Upstreams = []string{
+		setup.backend.URL + upstreamPath,
+	}
+	opts.Validate()
+
+	setup.proxy = NewOAuthProxy(opts, func(string) bool { return true })
+
+	return setup
+}
+
+func (f *fakeUpstreamSetup) close() {
+	f.backend.Close()
+}
+
+// Find out what RequestURI the proxy wound up serving for the requested one.
+func (f *fakeUpstreamSetup) translateRequest(path string) string {
+	rw := httptest.NewRecorder()
+	f.lastSeenURI = ""
+	f.lastSeenRequest = ""
+	f.lastSeenCode = 0
+	f.lastCode = 0
+	req, _ := http.NewRequest("GET", "http://localhost"+path, nil)
+	f.proxy.ServeHTTP(rw, req)
+	f.lastCode = rw.Code
+	return f.lastSeenURI
+}
+
+func TestNoSlashRequest(t *testing.T) {
```

Comment:
Table-driven can be useful, but I find it much easier to track down failing tests when there are line numbers and/or comments providing extra context. Plus, they are often just easier to read (IMHO).

Reply:
With table driven you can get the line numbers etc, give your test cases names and then it's pretty easy to see how the tests would work, if any case fails, the name will be in the failure message which then would give you the ability to look up the failed test case by name

Eg
```
FAIL TestTranslateRequest/Empty_Request_String
```

```
testCases := []struct{  
  name string
  requestString string
  expectedLastSeenURI string
  expectedLastSeenCode int
}{
  {
    name: "Empty Request String",
    requestString: "",
    expectedLastSeenURI: "",
    expectedLastSeenCode: 0,  
  },
  ...
}

for _, tc := range testCases {
  t.Run(tc.name, func(t *testing.T){
    fu := newFakeUpstreamSetup("")
    defer fu.close()

    fu.translateRequest(tc.requestString)
    assert.Equal(t, tc.expectedLastSeenURI, fu.lastSeenURI)
    assert.Equal(t, tc.expectedLastSeenCode, fu.lastSeenCode)
  }
}

```

Reply:
Yeah, I understand the request, I just don't think it makes much sense here.
1. The checks aren't all of exactly the same structure (some check .lastCode vs .lastSeenCode)
2. It involves unnecessarily creating the setup code for each sub-test which is:
   1. Slower
   2. Adds MORE boilerplate, because now you have to embed the "setup" URL/prefix for each test declaration.

Reply:
I think it's more important to improve maintainability here. Given what has been outlined, IMO shifting to table driven testing would help.